src/plugins/interactive_setup/server/plugin.ts

```diff
@@ -0,0 +1,155 @@
+    const { enabled } = this.#initializerContext.config.get<ConfigType>();
+
+    const isDev = this.#initializerContext.env.mode.dev;
+    const isConfigured =
+      core.elasticsearch.config.credentialsSpecified && core.elasticsearch.config.hosts.length;
+
+    /**
+     * env: dev -> enabled: false -> skip
+     * env: dev -> enabled: undefined -> skip
+     * env: dev -> enabled: true -> hold setup
+     * env: prod -> enabled: false -> skip
+     * env: prod -> enabled: undefined -> hold setup if not configured
+     * env: prod -> enabled: true -> hold setup if not configured
+     */
+    const shouldHoldSetupPhase = isDev
+      ? enabled === true
+      : enabled !== false
+      ? !isConfigured
+      : false;
+
+    if (!shouldHoldSetupPhase) {
+      return;
+    }
+
+    core.http.registerRoutes('', (prebootRouter) => {
+      let completeSetup: (result: { shouldReloadConfig: boolean }) => void;
+
+      prebootRouter.post(
+        {
+          path: '/api/preboot/setup',
+          validate: {
+            body: schema.oneOf([
+              schema.object({
+                hosts: schema.arrayOf(schema.string(), { minSize: 1 }),
+                apiKey: schema.string(),
+                caFingerprint: schema.string(),
+              }),
+              schema.object({
+                hosts: schema.arrayOf(schema.string(), { minSize: 1 }),
+                username: schema.string(),
+                password: schema.string(),
+                caFingerprint: schema.string(),
+              }),
+            ]),
+          },
+          options: { authRequired: false },
+        },
+        async (context, request, response) => {
+          if (!core.preboot.isSetupOnHold()) {
+            return response.badRequest();
+          }
+
+          const client = core.elasticsearch
+            .createClient('data', {
+              hosts: request.body.hosts.map((host) => `https://${host}`),
+              ssl: { verificationMode: 'none' },
+              // caFingerprint: request.body.caFingerprint,
+            })
+            .asScoped({
+              headers: {
+                authorization:
+                  'apiKey' in request.body
+                    ? `ApiKey ${btoa(request.body.apiKey)}`
+                    : `Basic ${btoa(`${request.body.username}:${request.body.password}`)}`,
+              },
+            });
+
+          try {
+            const { body } = await client.asCurrentUser.transport.request({
+              method: 'GET',
+              path: '/_security/enroll/kibana',
+            });
+
+            const configPath = isDev
+              ? this.#initializerContext.env.configs.find((path) => path.includes('dev'))
+              : this.#initializerContext.env.configs[0];
+
+            if (!configPath) {
+              return response.customError({ statusCode: 500, body: 'Cannot find config file.' });
+            }
+
+            await fs.appendFile(
+              configPath,
+              `
+
```

Comment:
__note:__ we also probably want to put some comment, like `This is auto-generated with blah-blah....`